File: 3789-maximize-subarrays-after-removing-one-conflicting-pair/maximize-subarrays-after-removing-one-conflicting-pair.java
class Solution {
    public long maxSubarrays(int n, int[][] conflictingPairs) {
        // Step 1: Build conflict list where conflicts[i] contains all values j such that (j, i) is a conflicting pair and j < i
        List<List<Integer>> conflicts = new ArrayList<>();
        for (int i = 0; i <= n; i++)
            conflicts.add(new ArrayList<>());

        for (int[] c : conflictingPairs) {
            int left = c[0], right = c[1];
            // Ensure left < right for consistency
            if (left > right) {
                int temp = left;
                left = right;
                right = temp;
            }
            // Register this pair as a conflict with 'right' depending on 'left'
            conflicts.get(right).add(left);
        }

        // Step 2: Prepare to compute total subarrays and restrict data
        long[] restrictRemoval = new long[n + 1]; // restrictRemoval[i]: gain in subarrays if we remove conflict involving i
        int leftMaxRestrict = 0, leftSecondMaxRestrict = 0; // track the 2 most restrictive positions (left ends of conflict pairs)
        long res = 0L;

        // Step 3: Iterate through the array from 1 to n
        for (int i = 1; i <= n; i++) {
            for (Integer ele : conflicts.get(i))
                // Update top 2 largest conflicting left ends for current i
                if (ele > leftMaxRestrict) {
                    leftSecondMaxRestrict = leftMaxRestrict;
                    leftMaxRestrict = ele;
                } else if (ele > leftSecondMaxRestrict)
                    leftSecondMaxRestrict = ele;

            // For index i, we can form (i - leftMaxRestrict) subarrays ending at i that do not contain any full conflicting pair
            res += (long) (i - leftMaxRestrict);

            // Store potential gain if we remove the conflict (leftMaxRestrict, i)
            restrictRemoval[leftMaxRestrict] += (long) (leftMaxRestrict - leftSecondMaxRestrict);
        }

        // Step 4: Find the best conflicting pair to remove (maximize gain in subarrays)
        long maxRemovalVal = 0L;
        for (int i = 1; i <= n; i++)
            maxRemovalVal = Math.max(maxRemovalVal, restrictRemoval[i]);

        // Step 5: Add the gain from removing the best conflicting pair
        res += maxRemovalVal;
        return res;
    }
}

File: 3789-maximize-subarrays-after-removing-one-conflicting-pair/maximize-subarrays-after-removing-one-conflicting-pair.py
class Solution:
    def maxSubarrays(self, n: int, conflictingPairs: List[List[int]]) -> int:
        conflicts = [[] for _ in range(n + 1)]

        # Build conflict map: conflicts[i] = all j < i that conflict with i
        for a, b in conflictingPairs:
            left, right = min(a, b), max(a, b)
            conflicts[right].append(left)

        restrict_removal = [0] * (n + 1)
        left_max_restrict = 0
        left_second_max_restrict = 0
        res = 0

        for i in range(1, n + 1):
            for ele in conflicts[i]:
                if ele > left_max_restrict:
                    left_second_max_restrict = left_max_restrict
                    left_max_restrict = ele
                elif ele > left_second_max_restrict:
                    left_second_max_restrict = ele

            res += i - left_max_restrict
            restrict_removal[left_max_restrict] += left_max_restrict - left_second_max_restrict

        max_removal_val = max(restrict_removal)
        return res + max_removal_val